test/replication-luatest/quorum_test.lua

```diff
@@ -0,0 +1,416 @@
+        if servers[i].net_box:eval('return box.info.id') ~= i then
+            t.helpers.retrying({timeout = 20}, function()
+               t.assert_equals(servers[i].net_box:eval(
+                   'return box.info.replication[' .. i .. '].upstream.status'),
+                   'follow',
+                   servers[i].alias .. ': this server does not follow others.')
+            end)
+        end
+    end
+end
+
+g.test_box_cfg_doesnt_return_before_all_replicas_are_not_configured = function()
+    -- Check that box.cfg() doesn't return until the instance
+    -- catches up with all configured replicas.
+    t.assert_equals(
+        quorum3.net_box:eval(
+            'return box.error.injection.set("ERRINJ_RELAY_TIMEOUT", 0.001)'),
+        'ok')
+    t.assert_equals(
+        quorum2.net_box:eval(
+            'return box.error.injection.set("ERRINJ_RELAY_TIMEOUT", 0.001)'),
+        'ok')
+    quorum1:stop()
+    t.helpers.retrying({timeout = 20}, function()
+        t.assert_not_equals(
+            quorum2.net_box:eval('return box.space.test.index.primary'), nil)
+    end)
+    quorum2.net_box:eval(
+        'for i = 1, ' .. COUNT .. ' do box.space.test:insert{i} end')
+    quorum2.net_box:eval("fiber = require('fiber')")
+    quorum2.net_box:eval('fiber.sleep(0.1)')
+    quorum1.args = {'0.1'}
+    start_servers({quorum1})
+    t.helpers.retrying({timeout = 20}, function()
+        t.assert_equals(
+            quorum1.net_box:eval('return box.space.test:count()'), COUNT)
+    end)
+    -- Rebootstrap one node of the cluster and check that others follow.
+    -- Note, due to ERRINJ_RELAY_TIMEOUT there is a substantial delay
+    -- between the moment the node starts listening and the moment it
+    -- completes bootstrap and subscribes. Other nodes will try and
+    -- fail to subscribe to the restarted node during this period.
+    -- This is OK - they have to retry until the bootstrap is complete.
+    local servers = {quorum1, quorum2, quorum3}
+    for _, server in ipairs(servers) do
+        t.assert_equals(server.net_box:eval('return box.snapshot()'), 'ok')
+    end
+end
+
+g.test_id_for_rebootstrapped_replica_with_removed_xlog = function()
+    quorum1:stop()
+    cleanup(quorum1.workdir)
+    fio.rmtree(quorum1.workdir)
+    fio.mktree(quorum1.workdir)
+    -- The rebootstrapped replica will be assigned id = 4,
+    -- because ids 1..3 are busy.
+    quorum1.args = {'0.1'}
+    start_servers({quorum1})
+
+    t.helpers.retrying({timeout = 20}, function()
+        t.assert_equals(
+            quorum1.net_box:eval('return box.space.test:count()'), COUNT)
+        t.assert_equals(
+            quorum2.net_box:eval(
+                'return box.info.replication[4].upstream.status'), 'follow')
+        t.assert(quorum3.net_box:eval('return box.info.replication ~= nil'))
+        t.assert_equals(
+            quorum3.net_box:eval(
+                'return box.info.replication[4].upstream.status'), 'follow')
+    end)
```

Comment:
Raw idea:

I may be biased, because I'm used to test-run, but here's my 2 cents.

Since there are a lot of evals in our tests, it would be nice to have some eval helper, which could
eval multiple commands at once, and return their results. And, maybe, compare them with reference?
I mean something like
```lua
quorum3:eval_and_cmp("return box.info.status", "running",
                     "return box.info.lsn",    10,
                     "return box.snapshot()",  true)
                        
```
This could shorten the tests quite a bit.

My design might be bad, but you got the idea, I guess.

Reply:
Also, maybe we can omit the `return`? eval('cmd') may already return the `cmd` result

Reply:
> My design might be bad, but you got the idea, I guess.

I think this is the wrong way, cause we want sequentially compare different fields/action results. This method is about comparing a few fields together. Some pros of assertion-based checks would be missed.

Reply:
[Raw idea] How about such API?

```lua
server_object.assert_equals('box.info.status', 'running')
```

1. Automatic 'return' seems useful.
2. We can add all (or at least most) luatest assertions to the server object this way.
3. It looks concise.

@rosik What do you think?

Reply:
Looks ambiguous. Dislike.

Don't try to put everything in test helpers. If you see some code repeats a lot in the test, just create an appropriate assertion locally. 

Here're some examples: [[1]](https://github.com/tarantool/cartridge/blob/75a64f183ca9013dc23333ae10303e0e305a1157/test/unit/roledeps_test.lua#L10-L16), [[2]](https://github.com/tarantool/cartridge/blob/75a64f183ca9013dc23333ae10303e0e305a1157/test/integration/pool_map_call_test.lua#L63-L84).

You can put it wherever you want: in a test function, in a test group, in repo helpers.